File: 프로그래머스/unrated/150368. 이모티콘 할인행사/이모티콘 할인행사.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;


vector<int> answer;
vector<vector<int>> users;
vector<int> emoticons;
vector<int> discountRates;
int n, m;
int arr[4] = {10, 20, 30, 40};
int EPS = 0;
int maxProfit = 0;

void dfs() {
    if(discountRates.size() == m){
        
        int eps = 0;
        int totalPrice = 0;
        for(auto user : users) {
            int userRate = user[0];
            int userLimit = user[1];
            int sum = 0;

            for(int i = 0 ; i < m ; i++) {
                if(discountRates[i] >= userRate) {
                    sum += emoticons[i] * (100 - discountRates[i]) / 100;
                } 
            }
            if(sum >= userLimit) {
                eps++;
            } else {
                totalPrice += sum;
            }
        }

        if(eps > EPS) {
            EPS = eps;
            maxProfit = totalPrice;
         } else if(eps == EPS && totalPrice > maxProfit) {
            maxProfit = totalPrice;
        }
        return;
    } 
    
    for(int i = 0 ; i < 4 ; i++) {
        discountRates.push_back(arr[i]);
        dfs();
        discountRates.pop_back();
    }
};

vector<int> solution(vector<vector<int>> _users, vector<int> _emoticons) {
    users = _users;
    emoticons = _emoticons;
    n = _users.size();
    m = _emoticons.size();
    
    dfs();
    
    answer.push_back(EPS);
    answer.push_back(maxProfit);
    return answer;
    
}

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.js
let users;
let emoticons;
let discountRates = [];
let answer = [0, 0];
 
function solution(_users, _emoticons) {
    users = _users;
    emoticons = _emoticons;
    dfs(0);
    return answer;
}

function dfs(idx) {
    if(idx === emoticons.length) {
        const [registerAmount, revenue] = getResult();
        
        if(registerAmount > answer[0]) {
            answer = [registerAmount, revenue];
        } else if(registerAmount === answer[0] && revenue > answer[1]) {
            answer = [registerAmount, revenue];
        }
        return;
    }
    
    for(let i = 10 ; i <= 40 ; i += 10) {
        discountRates.push(i);
        dfs(idx + 1);
        discountRates.pop();
    }
}

function getResult() {
    let curRegisterAmount = 0;
    let curRevenue = 0;
    
    users.forEach((user) => {
        const percentage = user[0];
        const limitPrice = user[1];
        
        const sum = emoticons.reduce((acc, price, idx) => {
            const curDiscountRates = discountRates[idx];
            
            if(curDiscountRates >= percentage) {
                acc += price * (100 - curDiscountRates) / 100;
            }
            
            return acc;
        }, 0);
        
        if (sum >= limitPrice) {
            curRegisterAmount++;
        } else {
            curRevenue += sum;
        }
    });
    
    return [curRegisterAmount, curRevenue];
}